MAINT: stats: undo `dirichlet` deprecation

mdhaber opened the pull request:
#### Reference issue
https://github.com/scipy/scipy/pull/16042#issuecomment-1137227051

#### What does this implement/fix?
This undoes the deprecation of `dirichlet` to allow for further discussion and adds a note to help users transition from `dirichlet` to `multivariate_beta`.

#### Additional information
We still intend to deprecate `dirichlet` or at least the `pdf`, `logpdf`, and `rvs` methods. It seems we need more time for discussion, though, and SciPy 1.9.0 will branch soon, so it is better to miss a release for the deprecation than to deprecate too soon.

tupui commented on scipy/stats/__init__.py:
```diff
@@ -161,7 +161,7 @@
    matrix_normal          -- Matrix normal distribution
    multivariate_beta      -- Multivariate beta distribution (Dirichlet)
-   dirichlet              -- Dirichlet (deprecated; see `multivariate_beta`)
+   dirichlet              -- Dirichlet (for new code, use `multivariate_beta`)
```
I had to remove this from the guide to pass the ref guide CI.

mdhaber replied:
But that is a problem with the refguide check, not the documentation. Shouldn't the check be fixed?

tupui replied:
I suppose we can argue with both. If we deprecate something, we don't want to show old things to people as they usually do not care about warnings until they break. Making things less discoverable can help further.

mdhaber replied:
With this PR, the `dirichlet` _distribution_ would not be deprecated; only the `pdf` and `logpdf` methods.

mdhaber commented:
Marking this as draft because we're going to try to change the `rvs` method of `multivariate_beta`.

mdhaber commented:
@oscarbenjamin would this PR resolve your concern? It would only raise the deprecation warning on calls to the `pdf` and `logpdf` methods of the distribution (and frozen distribution).

![image](https://user-images.githubusercontent.com/6570539/170351353-c4fd54f7-cc9f-4cf9-98e3-c39fd30a6b04.png)

For convenience, [here is a link](https://output.circle-artifacts.com/output/job/81943a17-ec52-4ef1-8209-91c860466064/artifacts/0/html-scipyorg/reference/generated/scipy.stats.dirichlet.html#scipy.stats.dirichlet) to the documentation.

tupui commented on scipy/stats/_multivariate.py:
```diff
@@ -1296,16 +1296,43 @@ def _lnB(alpha):
+        All other methods of `multivariate_beta` and `dirichlet` are identical;
+        use of other methods  does not need to be changed.
+
+        In SciPy 1.11.0, this deprecation warning will be removed and
+        `dirichlet` will become an alias for `multivariate_beta`.
```
I don't understand, are we deprecating the whole function or just two methods. The first line of the warning is misleading to me.

mdhaber replied:
This is not my idea. I'm trying to do something like what was suggested in https://github.com/scipy/scipy/pull/16042#issuecomment-1137227051.

It makes some sense to say that we are deprecating the _behavior_ of the `pdf` and `logpdf` methods. In 1.11, the behavior of those methods will change (to be the same as `multivariate_beta`). So at that point, the two names will be aliases of one another.

tupui replied:
Ok got it. But I am not sure I agree with doing something special here. I am never a fan of aliases like these. Users can have it on their end if they want to. IMO it's just confusing to have these in our codebase as it just increases the public API surface.

oscarbenjamin commented:
Thanks for opening this. Sorry if my other message came across as ranty. It wasn't meant to be. I just got carried away thinking about how to improve deprecation processes in general...

> However, this is still in flux: we also think that the unusual output shape of the `rvs` method should be fixed in `multivariate_beta`, so we will probably also raise a `DeprecationWarning` on calls to `rvs`. At that point, should we deprecate the individual methods, or deprecate the whole distribution?

I'm not sure I understand what else you expect to change here. I'm just coming at this from a high level so maybe this is obvious but... If there is to be a new better API called `multivariate_beta` then why does anything about `dirichlet` need to be changed (except maybe if some bug is fixed)?

Usually adding new API is considered to be the backwards compatible way of fixing changes that would break downstream code if applied to the old in-use API. The idea is you leave the old API sitting there, write the new API that does what you think is better and then you write in the docs that users should use the new API unless they need to support old versions.

Of course I understand the need to break compatibility sometimes but I think in this case a direct compatibility break (i.e. just fixing `dirichlet` directly) would make more sense to me than the confused messaging here:

1. The `dirichlet` class is broken so change your code to use `multivariate_beta`.
2. We need to warn you about this so that you change your code so that we can fix the `dirichlet` class in a future release.
3. But you shouldn't use `dirichlet` anyway even if we fix it later so change your code.
4. But we still need to fix it even though you shouldn't use it...

mdhaber commented:
> Usually adding new API is considered to be the backwards compatible way of fixing changes that would break downstream code if applied to the old in-use API. The idea is you leave the old API sitting there, write the new API that does what you think is better and then you write in the docs that users should use the new API unless they need to support old versions.

That is exactly what was done in gh-16042, right? We added the new API `multivariate_beta`, deprecated `dirichlet`, and the documentation pointed users of `dirichlet` to `multivariate_beta`.

> make more sense to me than the confused messaging here:

Ultimately, is the point of this part is that we should remove "In SciPy 1.11.0, this deprecation warning will be removed and `dirichlet` will become an alias for `multivariate_beta`."
Is that the part that made the message confusing?
Otherwise, the deprecation message I added to the documentation is almost identical to what you suggested in gh-16042. 

Please consider adding inline suggestions.

oscarbenjamin commented:
> That is exactly what was done in [gh-16042](https://github.com/scipy/scipy/pull/16042), right?

Not exactly. The old API has been changed so that it now emits a warning. Since I don't want that warning that API is now unusable to me. I might be wrong but it seems that you consider that the change from not emitting a warning to emitting a warning is not a break in compatibility but that's something that I would disagree with. I don't want my users to see irrelevant warnings that are not to do with *their* code.

tupui commented:
I see your point but since SciPy is also used by end users we need to "warn" users that the function they are using is going to disappear. This is exactly what deprecation warnings are for and it's normal practice. You can filter this warning in your code so that your users don't see it as you took care of it.

But the warning exactly served it's purpose as you have seen it and are thinking about how to evolve your code for future compatibility.

oscarbenjamin commented:
Wait but is the function going to disappear? I thought the intention was for it to remain but in future `dirichlet` is to become an alias for `multivariate_beta` which would have the effect of changing its `pdf` and `logpdf` methods to effectively transpose their inputs. The warning as I understood it was that in future the behaviour of those two methods is expected to change.

My first question though is why that change needs to happen at all rather than how to warn about it. The new API `multivariate_beta` already solves the problems with `dirichlet`. The long term hope is that downstream code should switch to using that rather than using `dirichlet`. So, doesn't it make sense just to leave `dirichlet` as it was (apart from refactoring to share code with `multivariate_beta`)?

I think if I was going down this path I would have made `dirichlet` subclass `multivariate_beta` just to override those two methods. Then there isn't really any maintenance burden for leaving `dirichlet` effectively unchanged. The deprecation here just seems to be making unnecessary work for SciPy, downstream libraries and ultimately users without really offering any benefit (that I can see).

tupui commented:
The current plan is to completely remove the function and have an alias (which I disagree with).

> The long term hope is that downstream code should switch to using that rather than using dirichlet. So, doesn't it make sense just to leave dirichlet as it was (apart from refactoring to share code with multivariate_beta)?

If the goal is to transition to `multivariate_beta`, then no in the end we just want to have this function in the code. Having both is unnecessary maintenance and misleading to new users as they can still find the old function and use it. People don't update unless code is broken.

oscarbenjamin commented:
> Having 2 APIs for the same thing is a maintenance burden, always.

I agree but bear in mind that unless `dirichlet` is immediately removed (which no one is suggesting) then the work needed to make both APIs work right now has to be done in any case. I would approach this by refactoring so that the additional effort needed to keep `dirichlet` in place (at least as long as whatever deprecation period) is minimised. Then at that point the maintenance burden vs backwards compatibility can be considered but I wouldn't start issuing warnings to users until the new API is established in released code.

tupui commented:
> > Having 2 APIs for the same thing is a maintenance burden, always.
> 
> I agree but bear in mind that unless `dirichlet` is immediately removed (which no one is suggesting) then the work needed to make both APIs work right now has to be done in any case.

Yes and this is not the current state. The current state is that when you use `dirichlet`, it raises a deprecation warning. The class will only be removed in 2 versions.

> I would approach this by refactoring so that the additional effort needed to keep `dirichlet` in place (at least as long as whatever deprecation period) is minimised. Then at that point the maintenance burden vs backwards compatibility can be considered but I wouldn't start issuing warnings to users until the new API is established in released code.

My understanding is that the new API (`multivariate_beta`) is established. The guidance Matt provided was to help you find a way in your code to handle the deprecation if you needed to have the output formatted the way `dirichlet` is doing it.

Yes it's always painful to support multiple versions and we have a few hooks in SciPy as well to handle multiple versions of NumPy transparently.

oscarbenjamin commented:
> My understanding is that the new API (`multivariate_beta`) is established

My definition of "established" is something like: the API was released long enough ago that no one really needs to think twice about changing their code to use it. If new API comes into existence in the same release that the old API is deprecated then it's too soon to just update your code to use the new API.

> Yes it's always painful to support multiple versions

To be clear it isn't a problem for SymPy to update its code here. Rather the problem is that even if we put out a new release of SymPy tomorrow there will be cases in the future where someone has an older version of SymPy with a newer version of SciPy and will see the warnings. Ideally our processes for these things would make it possible to ensure that users rarely see warnings except if there's a problem with their own code. That's not possible though if the visible deprecation warnings and the new API both appear in the same release.

Other than that this kind of thing is perfectly manageable for SymPy and with a small amount of regular CI maintenance we can smooth over these sorts of changes. Mainly I'm questioning the longer term approach here on behalf of users and less well-maintained libraries. The cost here is that their code ends up broken in some way (if only because it emits warnings). The benefit to counterbalance that is some reduction in maintenance cost but I think that the maintenance cost can be made very small meaning that the benefit of breaking compatibility is also small.

tupui commented:
I understand better your point, thank you for explaining.

> Rather the problem is that even if we put out a new release of SymPy tomorrow there will be cases in the future where someone has an older version of SymPy with a newer version of SciPy and will see the warnings.

This is IMO a user problem. They either have to update SciPy or filter the warning themselves.

Otherwise this would mean:

We are on version `n`. If we create a new function that should deprecate another one, we need to wait for at least `n+4` (2 years, NEP29 but some project go even further and support even more versions) until we can put a deprecation warning on the old one. This way in SymPy for instance you would be able to update the minimal version following NEP29 and users updating to a new version would not see any warnings. And only in `n+6` we can remove the old function (so 3 years).

IMHO this is too much.

tylerjereddy commented:
@oscarbenjamin Concisely, do you prefer what this PR does to what we merged a day or two ago?

@mdhaber @tupui so, this is basically switching to a soft "documented deprecation" instead of a "proper deprecation" to be more conservative for now?

mdhaber commented:
@oscarbenjamin We may actually want to deprecate all of `dirichlet` in the near future, so rather than doing either, this PR now reverts the deprecation of `dirichlet` entirely to leave more time for discussion (before or after 1.9.0 branches). 

One issue is about whether we are deprecating (or should deprecate) all of `dirichlet` or just the `pdf`, `logpdf`, and `rvs` methods (which will be changed in `multivariate_beta` shortly). IIUC, your view is that if only the behavior of certain methods is changing, only those methods should emit a deprecation warning. Instead, I had chosen to emit deprecation warnings from all methods of the `dirichlet` object because after the deprecation, it would no longer be an instance of the same `dirichlet_gen` class. But this issue may become moot if we decide not to offer `dirichlet` as an alias after all. 

You also gave suggestions deprecation procedures, e.g. not deprecating in the same release as the new API and maybe emitting a `PendingDeprecationWarning` before beginning to emit a `DeprecationWarning`. For this, I'd suggest joining the [scipy-dev mailing list](https://mail.python.org/mailman3/lists/scipy-dev.python.org/) and posting this as a suggestion to the group. I don't think we broke standard practice with gh-16042, so I'd appreciate it if you would phrase the comment as a general suggestion rather than singling out gh-16042. (gh-16042 closed an issue that had been reported at least twice for `dirichlet` in particular (gh-4984, gh-6006), and it's probably fear of problems like this that nobody wanted to touch it for the past seven years!) We did our best to consider BC when writing gh-16042, considering that it would have been a one-line change otherwise, but we failed to anticipate all concerns. Another reason to join the mailing list is that you can help us catch this sort of thing when an [email is sent](https://mail.python.org/archives/list/scipy-dev@python.org/thread/OL33GITKJJLIVEZWPLG4LECZKDEUUILN/). Thanks!

mdhaber commented:
@tylerjereddy yes, that's the intent.

oscarbenjamin commented:
> @oscarbenjamin Concisely, do you prefer what this PR does to what we merged a day or two ago?

Yes I think this PR is an improvement to the messaging and the scope of the deprecation... except it depends what the ultimate plan is. It is *still* unclear to me what the expected future of the `dirichlet` API is. As a general principle I don't think users should see any warnings or suggestions to change their code before a clear plan is in place. Of course the plan can be changed later if needed but there should be *some* working plan and it should be documented somewhere.

If you look at the OP of #16042 then you can see "After the deprecation cycle, we can make dirichlet an alias for multivariate_beta if desired". If you look above then you can see "we need to "warn" users that the function they are using is going to disappear". Also above "We may actually want to deprecate all of dirichlet in the near future". Which is it?

What is the *expected ultimate end state* here? That should be spelled out before asking anyone else to change their code.

Separately from that I don't see any strong reason given anywhere for not simply leaving the `dirichlet` API as it was *given that the fixes are going into a new API*. As far as I can tell the improvements in #16042 can be achieved (more) easily without any observable changes to `dirichlet`. A simple refactor could reduce `dirichlet` to a handful of code that just overrides a couple of methods of `multivariate_beta`

Perhaps this will help to understand how users might encounter this in future: I saw the deprecation message in sympy's CI (which I had just changed to use scipy main). The message said to use `multivariate_beta` so I opened ipython and give it a go but no dice because my local scipy didn't have the function. A web search for "scipy multivariate_beta" lead nowhere. So I went to the scipy github and searched for `multivariate_beta` in the issues: no open or closed issues. Then searching in (closed) PRs I found a single PR #16042. While I did figure that out I think it's probably beyond most users who might encounter the message to find their way there. Then when I did get there it was still unclear to me what had been changed and why and exactly how I should change my code in response.

I commented in #16042 and there has been discussion here since but after reading all comments again I *still* don't know what the plan is from scipy's perspective. This PR improves the messaging by making it clearer what has been changed already in main but I am still unclear what the plan is for `dirichlet` in future. It's hard to comment on the suitability of this PR without knowing that plan.

mdhaber commented:
The original PR gh-16042 kicked off the following plan (discussed in an [email to the mailing list](https://mail.python.org/archives/list/scipy-dev@python.org/thread/OL33GITKJJLIVEZWPLG4LECZKDEUUILN/)) in motion.

1. Introduce `multivariate_beta`, and deprecate all of `dirichlet`, warning users of all methods of the deprecation and advising them to transition to `multivariate_beta`. 
After the deprecation cycle, we would remove the `dirichlet_gen` class an no longer instantiate the `dirichlet` object as:
https://github.com/scipy/scipy/blob/26eb83029dd13ac24b2d3cdc5946b34010dc934c/scipy/stats/_multivariate.py#L1578

2. Instead, after the deprecation cycle, it would be `dirichlet = multivariate_beta`, where:
https://github.com/scipy/scipy/blob/26eb83029dd13ac24b2d3cdc5946b34010dc934c/scipy/stats/_multivariate.py#L1729

After the concerns over the appropriateness of #1 and the necessity of #2 were raised, this PR puts that plan on hold, instead only introducing `multivariate_beta`. The note advising users of the `pdf` and `logpdf` methods of `dirichlet` to adopt `multivariate_beta` is presented in a style similar to the one NumPy uses to recommend that users transition from `assert_almost_equal` to `assert_allclose`. In any case, users of the `pdf` and `logpdf` methods of `dirichlet` will not be able to leave their code untouched indefinitely - they will need to transition to the convention of `multivariate_beta` at some point.

>Separately from that I don't see any strong reason given anywhere for not simply leaving the `dirichlet` API as it was given that the fixes are going into a new API. 

The reason is that we don't think it is appropriate to have two nearly-identical APIs for the same distribution, one which is inconsistent with itself and its own RVS method (`dirichlet`), and one which is correct (`multivariate_beta`).

> As far as I can tell the improvements in https://github.com/scipy/scipy/pull/16042 can be achieved (more) easily without any observable changes to `dirichlet`. A simple refactor could reduce `dirichlet` to a handful of code that just overrides a couple of methods of multivariate_beta.

There is nearly minimal code duplication between `dirichlet` and `multivariate_beta` as-is. The diff in gh-16042 is bloated because of the documentation for the two classes (`dirichlet_gen` and `multivariate_beta_gen`) is tailored to the distribution. As you suggest, there is only one essential override that distinguishes the two behaviors:
https://github.com/scipy/scipy/blob/26eb83029dd13ac24b2d3cdc5946b34010dc934c/scipy/stats/_multivariate.py#L1427-L1429
vs
https://github.com/scipy/scipy/blob/26eb83029dd13ac24b2d3cdc5946b34010dc934c/scipy/stats/_multivariate.py#L1718-L1720

The rest is some boilerplate for freezing the duplications. I've mentioned elsewhere that there should be provisions for this inherited from the superclass `rv_multivariate`, but I don't think there's any unusual code duplication there.

In any case, the need to change `dirichet`'s behavior is not related to maintenance burden. We just think it's wrong to have a distribution class that is different from most of the rest and for which the `pdf` method doesn't accept the output of `rvs`.

tylerjereddy commented:
Sounds good, I didn't really see anything out of line in the original deprecation either. I also see this happen to SciPy proper when NumPy deprecates stuff, and sometimes those are tentative adjustments they make on `main` as well, so not sure we're so unusual here in signaling that change is coming without a formal spec of the final product or a decisive conclusion on whether an alias should remain for a long time.

I think Ralf tends to have clear/decisive views on this stuff as i.e., the author of https://numpy.org/neps/nep-0023-backwards-compatibility.html (etc.).

mdhaber commented:
Yeah. Maybe I should just remove the statement that there will be an alias, since that seems to be the source of much of the uncertainty. We can always add it if it is desired, but we don't _have_ to guarantee it now.

oscarbenjamin commented:
This PR looks good to me. It improves the messaging in the docs for how to change code that uses `dirichlet` and as I understand it removes any warnings from `dirichlet`. If there isn't yet an agreement on the ultimate plan then I think that not emitting any warnings is the right move at this stage.

The only thing I would potentially change here is reversing the order of inheritance so that it is `dirichlet` that inherits from and overrides the behaviour of `multivariate_beta` rather than the other way round. It's a minor point but if the plan is that `multivariate_beta` is the preferred future API then I would implement that directly and have `dirichlet` as the minimised alternate implementation. That way makes the implementation of the preferred `multivariate_beta` simpler to understand and makes it clearer in the code that `dirichlet` only exists for backwards compat and could simply be deleted in future without needing to refactor `multivariate_beta` again.

If in future it is decided to make `dirichlet` an alias for `multivariate_beta` then what that really means is changing the behaviour of its `log` and `logpdf` methods so per NEP 23 emitting `FutureWarning` from those methods would make more sense then `DeprecationWarning`. On the other hand if it is agreed now to (eventually) remove `dirichlet` altogether then I would go with `PendingDeprecationWarning` for all of `dirichlet`'s methods right now. There doesn't need to be any delay in introducing `PendingDeprecationWarning` since it's only for CI and is not seen by users. Getting that warning out early gives downstream libraries maximum time to adapt before any user visible warnings appear.

mdhaber commented:
We're reverting the addition of `multivariate_beta` for other reasons. (Other subtle issues in `dirichlet` were deeper than previously thought, and we couldn't fix all of them in `multivariate_beta` in time.) So closing this.
